frontend/src/utils/account-send.ts

```diff
@@ -0,0 +1,238 @@
+  const data = buildAccountDataForEncryption({ recipientAddress, advancedMode, usePublicKeyChecked, pubKey });
+  const encryptedData = data.xor(encryptionKey);
+  return hexZeroPad(encryptedData.toHexString(), 32);
+};
+
+export const decryptData = (accountSendCiphertext: string, keyData: KeyData) => {
+  const { viewingPrivateKey, encryptionCount } = keyData;
+
+  assertValidHexString(viewingPrivateKey, 32, 'Invalid viewing key');
+  assertValidHexString(accountSendCiphertext, 32, 'Invalid ciphertext');
+  assertValidEncryptionCount(encryptionCount, 'Invalid count for decryption');
+
+  const encryptionCountHex = hexZeroPad(keyData.encryptionCount.toHexString(), 32);
+  const encryptionKey = keccak256(`${viewingPrivateKey}${encryptionCountHex.slice(2)}`);
+
+  const decryptedData = BigNumber.from(accountSendCiphertext).xor(encryptionKey);
+  const hexData = hexZeroPad(decryptedData.toHexString(), 32);
+
+  const partialPubKey = hexData.slice(44);
+  const advancedMode = hexData.slice(42, 43);
+  const usePublicKeyChecked = hexData.slice(43, 44);
+  return {
+    advancedMode: advancedMode === '1' ? true : false,
+    usePublicKeyChecked: usePublicKeyChecked === '1' ? true : false,
+    address: getAddress(hexData.slice(0, 42)),
+    pubKey: `0x99${partialPubKey}` as PartialPublicKey,
+  };
+};
+
+export const storeSend = async ({
+  chainId,
+  provider,
+  viewingPrivateKey,
+  unencryptedAccountSendData,
+  accountDataToEncrypt,
+}: StoreSendArgs) => {
+  const { amount, tokenAddress, txHash, senderAddress } = unencryptedAccountSendData;
+  const { recipientAddress, advancedMode, usePublicKeyChecked, pubKey } = accountDataToEncrypt;
+
+  assertValidAddress(senderAddress, 'Invalid sender address');
+  assertValidAddress(tokenAddress, 'Invalid token address');
+  assertValidHexString(viewingPrivateKey, 32, 'Invalid viewing key');
+  assertValidHexString(txHash, 32, 'Transaction hash');
+  assertValidPublicKey(pubKey);
+
+  // Send history is scoped by chain
+  const localStorageKey = `${LOCALFORAGE_ACCOUNT_SEND_KEY_PREFIX}-${senderAddress}-${chainId}`;
+  let count = (await localforage.getItem(
+    `${LOCALFORAGE_ACCOUNT_SEND_KEY_PREFIX}-count-${senderAddress}-${chainId}`
+  )) as string;
+  let values = ((await localforage.getItem(localStorageKey)) as AccountSendData[]) || [];
+
+  // If there is no count or the length of values is 0
+  // then someone cleared localstorage data,
+  // and we need to reset the count.
+  if (!count || values.length === 0) {
+    count = new RandomNumber().value.toString();
+
+    await localforage.setItem(
+      `${LOCALFORAGE_ACCOUNT_SEND_KEY_PREFIX}-count-${senderAddress}-${chainId}`,
+      count.toString()
+    );
+    // Clear values because they cannot be decrypted if the original count is missing.
+    await localforage.setItem(localStorageKey, []);
+    values = [];
+  }
+
+  const checksummedRecipientAddress = await toAddress(recipientAddress, provider);
```

Comment:
This var name is called `checksummedRecipientAddress` but the `toAddress` can return an ENS/CNS name. I think we want to use the ethersjs `getAddress` method here instead?

Reply:
I don't think that is the case. Based on the comment and implementation I am seeing it will return a checksummed address.

https://github.com/ScopeLift/umbra-protocol/blob/master/frontend/src/utils/address.ts#L66

https://github.com/ScopeLift/umbra-protocol/blob/f82e73ba28e98fac5f28d0320fb6593eadbf1107/umbra-js/src/utils/utils.ts#L177

Reply:
Ah whoops yep you're right